corgi/tasks/pulp.py

```diff
@@ -0,0 +1,63 @@
+import logging
+
+from celery_singleton import Singleton
+
+from config.celery import app
+from corgi.collectors.pulp import Pulp
+from corgi.core.models import (
+    Channel,
+    ProductComponentRelation,
+    ProductStream,
+    ProductVariant,
+)
+from corgi.tasks.brew import fetch_modular_builds
+from corgi.tasks.common import RETRY_KWARGS, RETRYABLE_ERRORS, _create_relations
+
+logger = logging.getLogger(__name__)
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+def setup_pulp_relations() -> None:
+    logger.info("Setting up CDN repo relations for all Channels")
+    for channel in Channel.objects.filter(type=Channel.Type.CDN_REPO):
+        for pv_ofuri in channel.product_variants:
+            pv = ProductVariant.objects.get(ofuri=pv_ofuri)
+            slow_setup_pulp_rpm_relations.delay(channel.name, pv.name)
+            slow_setup_pulp_module_relations.delay(channel.name, pv.name)
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+def slow_setup_pulp_rpm_relations(channel, variant):
+    srpm_build_ids = Pulp().get_rpm_data(channel)
+    no_of_relations = _create_relations(
+        srpm_build_ids, channel, variant, ProductComponentRelation.Type.CDN_REPO
+    )
+    if no_of_relations > 0:
+        logger.info("Created %s new relations for SRPMs in %s", no_of_relations, channel)
+
+
+@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
+def slow_setup_pulp_module_relations(channel, variant):
+    module_build_ids = Pulp().get_module_data(channel)
+    no_of_relations = _create_relations(
+        module_build_ids, channel, variant, ProductComponentRelation.Type.CDN_REPO
+    )
+    if no_of_relations > 0:
+        logger.info("Created %s new relations for rhel_modules in %s", no_of_relations, channel)
+
+
+def get_builds_by_cdn_repo(stream_names: list[str]):
+    logger.info("Called save cdn repo with streams %s", stream_names)
+    build_ids = set()
+    for ps in ProductStream.objects.filter(name__in=stream_names):
+        for pv in ProductVariant.objects.filter(name__in=ps.product_variants):
+            relations_query = (
+                ProductComponentRelation.objects.filter(
+                    product_ref=pv.name,
+                    type=ProductComponentRelation.Type.CDN_REPO,
+                )
+                .values_list("build_id", flat=True)
+                .distinct()
+            )
+            build_ids.update(fetch_modular_builds(relations_query))
```

Comment:
This looks like a bug - fetch_modular_builds() returns a list of ints, but the other build IDs here might be strings.

Reply:
fetch_modular_builds is also called by loadcomposedata module. We're still left with the cast to int there, but I think that's better than the alternative, which is to change the build_id parameter of slow_fetch_brew_build also to a str.

I have updated this as you suggested though.

Reply:
Now I just cast the build_id to an int when calling slow_fetch_brew_ build from fetch_modular_builds.